### PR TITLE
Fix join expression pushdown coercion

### PR DESCRIPTION
**Pull request by arhimondr**

### CODE REVIEW

File: presto-tests/src/main/java/com/facebook/presto/tests/AbstractTestQueries.java [hunk 1]
```diff
@@ -2077,6 +2078,12 @@ public void testJoinCriteriaCoercion()
                 "SELECT * FROM (VALUES (1, 2)) x (a, b) JOIN (VALUES (SMALLINT '1', SMALLINT '3')) y (a, b) ON x.a = y.a",
                 "VALUES (1, 2, 1, 3)"
         );
+
+        // short decimal, long decimal
+        assertQuery(
+                format("SELECT * FROM (VALUES (CAST(1 AS DECIMAL(%1$d,0)), 2)) x (a, b) , (VALUES (CAST(0 AS DECIMAL(%1$d,0)), SMALLINT '3')) y (a, b) WHERE x.a = y.a + 1", MAX_SHORT_PRECISION),
```
Comment (kokosing) [inline presto-tests/src/main/java/com/facebook/presto/tests/AbstractTestQueries.java]: can please add a test for: 
```
SELECT * FROM (VALUES (CAST(1 AS DECIMAL(%1$d,0)), 2)) x (a, b) , (VALUES (CAST(0 AS DECIMAL(%1$d,0)), SMALLINT '3')) y (a, b) ON x.a = y.a + 1;

SELECT * FROM (VALUES (CAST(1 AS DECIMAL(%1$d,0)), 2)) x (a, b) LEFT JOIN (VALUES (CAST(0 AS DECIMAL(%1$d,0)), SMALLINT '3')) y (a, b) ON x.a = y.a + 1;
```

Comment (kokosing) [inline presto-tests/src/main/java/com/facebook/presto/tests/AbstractTestQueries.java]: Additionally can you please format query to make it easier to read. Maybe:
```
SELECT * FROM 
    (VALUES (CAST(1 AS DECIMAL(%1$d,0)), 2)) x (a, b) , 
    (VALUES (CAST(0 AS DECIMAL(%1$d,0)), SMALLINT '3')) y (a, b) 
WHERE x.a = y.a + 1
```

File: presto-tests/src/main/java/com/facebook/presto/tests/AbstractTestQueries.java [hunk 2]
```diff
@@ -2077,6 +2078,12 @@ public void testJoinCriteriaCoercion()
                 "SELECT * FROM (VALUES (1, 2)) x (a, b) JOIN (VALUES (SMALLINT '1', SMALLINT '3')) y (a, b) ON x.a = y.a",
                 "VALUES (1, 2, 1, 3)"
         );
+
+        // short decimal, long decimal
+        assertQuery(
```
Comment (arhimondr) [inline presto-tests/src/main/java/com/facebook/presto/tests/AbstractTestQueries.java]: This is not a case. I was able to reproduce it only in expression pushdown case.

**Comment by martint** [comment]
Merged, thanks!

**Pull request closed without merging**